### DISCUSSION TITLE
Mesh adaptivity problem for multi-scale grain growth simulation

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#19275] [Q&A Meshing] [answered]
Dear MOOSE experts,

Recently, I used the bulit-in grain growth model in moose to conduct multi-scale bicrystal simulation, where the grain size is 20μm and the grain boudary (GB) width is 15nm. The model mesh diagram at the time step 3 and 7 are as follow,

![image](https://user-images.githubusercontent.com/56183775/139850880-fb52a9b8-8a8d-4160-9079-3a698aca472e.png)
![image](https://user-images.githubusercontent.com/56183775/139850930-b5c3f650-bed4-4554-af26-b3ed06429937.png)

However, we found that there are some problems with the subdivided grid after using the mesh adaptive technology. For example, the grid in the green box is finer than the grid in the bule box at the Time Step 3, as show in Fig.1. According to our understanding from the bicrystal grain growth simulation in the nanometer scale, the grid at the GB should be of uniform size. This problm is more obvious in the yellow box in Fig.2, and the change of gr0 at the GB is shown in the figure below,

![image](https://user-images.githubusercontent.com/56183775/139850962-15c5776e-b776-4eca-bc9a-fa2b1f9d60b7.png)

Moreover, this is the code for the mesh adaptivity technology I used,
```bash
  [./Adaptivity]
    initial_adaptivity = 7
    cycles_per_step = 2 # The number of adaptivity cycles per step
    refine_fraction = 0.5 # The fraction of elements or error to refine.
    coarsen_fraction = 0.05
    max_h_level = 10
  [../]
```
In general, our question is why the grid adaptation is so strange in the process of GB migration. How to better use the mesh adaptive technology in moose to realize such a multi-scale grain growth simulation? 

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by GiudGiud** [comment, marked as answer]
Hello

What did you use as the Indicator for this Adaptivity?
That will determine how refinement decisions are made. 

Guillaume

**Reply by PengWei97** [reply]
I just referred to the mesh adaptivity part in [grain_growth_2D_graintracker.i](https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/examples/grain_growth/grain_growth_2D_graintracker.i), and I did  not specify the Indicator and marker. So I don't know how the refinement is determined from the *.i file.

Wei

**Reply by GiudGiud** [reply]
I think this is the default error estimator (eg indicator of where to refine)
https://libmesh.github.io/doxygen/classlibMesh_1_1KellyErrorEstimator.html

You should look at these pages and choose the one you want to use:
https://mooseframework.inl.gov/syntax/Adaptivity/Indicators/index.html

**Reply by PengWei97** [reply]
Yeah, I modified my input file, [GG_Custom_Indicators.i](https://github.com/PengWei97/monkey/blob/main/problem/adaptivity_mesh/GG_Custom_Indicators.i) based on the webpage you shared and GrandPotentialPlanarGrowth.i. The code that contains the adaptation is as, 
```bash
[Adaptivity]
 initial_steps = ${my_number_adaptivity}
 max_h_level = ${my_number_adaptivity}
 initial_marker = err_eta
 marker = err_bnds
[./Markers]
   [./err_eta]
     type = ErrorFractionMarker
     coarsen = 0.3
     refine = 0.5
     indicator = ind_eta
   [../]
   [./err_bnds]
     type = ErrorFractionMarker
     coarsen = 0.3
     refine = 0.5
     indicator = ind_bnds
   [../]
 [../]
 [./Indicators]
   [./ind_eta]
     type = GradientJumpIndicator
     variable = gr0
    [../]
    [./ind_bnds]
      type = GradientJumpIndicator
      variable = bnds
   [../]
 [../]
[]
```
In addition, I also created the [GG_default_adaptive.i ](https://github.com/PengWei97/monkey/blob/main/problem/adaptivity_mesh/GG_default_adaptive.i)using the default adaptive technolohy. The code that contains the adaptation is as follows,
```bash
  [./Adaptivity]
    initial_adaptivity = ${my_number_adaptivity} # 5
    cycles_per_step = 2 # The number of adaptivity cycles per step
    refine_fraction = 0.5 # The fraction of elements or error to refine.
    coarsen_fraction = 0.05
    max_h_level = 8
  [../]
[]
```
Then, the mesh results of the two adaptive mesh setting are shown below,
![image](https://user-images.githubusercontent.com/56183775/140469852-6087c1a2-2c00-4fdd-afd3-ca2a53f435dc.png)

By comparing the two adaptive grid methods, I think that the custom adaptive grid division (fig.1) is not as perfect as the default grid adaptive method (fig.2). Although the first calculation only took 4.29 minutes, the second calculation took 15.45 minutes. Moreover, the circular grian-time curve of the two simultion outputs is the same,
![image](https://user-images.githubusercontent.com/56183775/140471392-edefadc1-8877-4e9c-a21e-64ff7953dc24.png)

So my final question is how to modify my GG_Custom_Indicators.i to make the quality of the adaptive grid better without spending as much time as GG_default_adaptive.i.

Wei

**Reply by GiudGiud** [reply]
are the number of elements similar in both runs?
To spend less time you usually want to limit the number of elements. So run less adaptivity steps and set a lower max refinement. 

You will also want to turn on the perf_graph = true in Outputs to see what is actually using the simulation time.

**Reply by PengWei97** [reply]
The number of grids initially set is the same, but the first uses bnds as indicators, and the second uses default indicators. However, the first mesh adaptive effect is not as good as the second one, so I would like to ask how to better set Indicators to improve the mesh quality after mesh adaptive division.

**Reply by GiudGiud** [reply]
this is very problem dependent.
You need to determine what you consider the "error" to be, and use that as a criteria for refinement. 

I was saying that the two refinement results dont seem to have the same number of elements, so that s where the runtime difference comes from. If you want the "better" refinement to be cheaper, you ll have to limit the number of levels / tune the parameters

**Reply by PengWei97** [reply]
OK, Thank you for your timely response!

Perhaps, I need to further modify my indicators block and control the number of elements within an acceptable range.
